server/lib/actors/hot.go

```diff
@@ -0,0 +1,741 @@
+type ByVote struct {
+	l *hotlist
+}
+func (by *ByVote) Len() int {
+	return by.l.Len()
+}
+func (by *ByVote) Swap(i, j int) {
+	by.l.Swap(i, j)
+}
+func (by *ByVote) Less(i, j int) bool {
+	// > for sorting decreasing order
+	e := by.l.entries[i]
+	return by.l.entries[j].lessThanByVote(e.Vote, e.Id)
+}
+
+
+//
+// hothotbucket
+//
+// hothotbucket represents total and per-parent hotlists in certain terms (day/year/week...)
+type hotbucket struct {
+	lastUpdate time.Time
+	total *hotlist
+	groups map[proto.UUID]*hotlist
+	idParentMap map[proto.UUID]proto.UUID
+}
+
+func NewHotBucket(size int) *hotbucket {
+	return &hotbucket {
+		total: NewHotList(size),
+		groups: make(map[proto.UUID]*hotlist),
+		idParentMap: make(map[proto.UUID]proto.UUID),
+	}
+}
+
+func NewHotBucketWithLastUpdate(size int, lastUpdate time.Time) *hotbucket {
+	hb := NewHotBucket(size)
+	hb.lastUpdate = lastUpdate
+	return hb
+}
+
+//addResults adds []FetchResult to bucket
+func (b *hotbucket) addResults(res []FetchResult) *hotbucket {
+	return b.addResultsWithFilter(res, nil)
+}
+
+func (b *hotbucket) addResultsWithFilter(res []FetchResult, filter func (e *HotEntry) bool) *hotbucket {
+	for _, r := range res {
+		e := &HotEntry{ Id: r.id, Score: r.score, Vote: r.vote }
+		if filter != nil && !filter(e) {
+			continue
+		}
+		b.addEntry(e, r.parent)
+	}
+	return b
+}
+
+func (b *hotbucket) addEntry(e *HotEntry, parent proto.UUID) {
+	b.total.addEntry(*e)
+	b.idParentMap[e.Id] = parent
+	l, ok := b.groups[parent]
+	if !ok {
+		l = NewHotList(PER_CHANNEL_HOTBUCKET_SIZE)
+		b.groups[parent] = l
+	}
+	l.addEntry(*e)
+}
+
+//filter filters hotentry with given filter function
+func (b *hotbucket) filter(filt func (e *HotEntry) bool) *hotbucket {
+	br := NewHotBucketWithLastUpdate(len(b.total.entries), b.lastUpdate)
+	for _, e := range b.total.entries {
+		if filt(&e) {
+			br.addEntry(&e, b.idParentMap[e.Id])
+		}
+	}
+	return br
+}
+
+//add adds another hotbucket. delta can be nil
+func (b *hotbucket) add(delta *hotbucket) *hotbucket {
+	if delta != nil {
+		b.total.addEntries(delta.total.entries)
+		for parent, l := range delta.groups {
+			cl, ok := b.groups[parent]
+			if ok {
+				cl.addEntries(l.entries)
+			} else {
+				b.groups[parent] = cl.dup()
+			}
+		}
+	}
+	return b
+}
+
+func (b *hotbucket) sort(btyp proto.TopicListRequest_BucketType) *hotbucket {
+	b.total.sort(btyp)
+	for _, l := range b.groups {
+		l.sort(btyp)
+	}
+	return b
+}
+
+//truncate bucket with specified size.
+func (b *hotbucket) truncate(size int) *hotbucket {
+	if size > len(b.total.entries) {
+		size = len(b.total.entries)
+	}
+	br := NewHotBucketWithLastUpdate(len(b.total.entries), b.lastUpdate)
+	for _, e := range b.total.entries[0:size] {
+		br.addEntry(&e, b.idParentMap[e.Id])
+	}
+	br.lastUpdate = b.lastUpdate
+	return br
+}
+
+func (b *hotbucket) rangeOf(ids []proto.UUID, btyp proto.TopicListRequest_BucketType, 
+	startScore int32, startId proto.UUID, count uint32) ([]HotEntry, error) {
```

Comment:
__[gometalinter]__
* [gotype] undeclared name: proto
* [golint] method parameter startId should be startID